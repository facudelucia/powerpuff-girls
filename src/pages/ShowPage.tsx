import React from 'react';
import ShowDetails from '../components/ShowDetails';
import styled from 'styled-components';


const PageContainer = styled.div`
  padding: 20px;

  @media (max-width: 768px) {
    padding: 15px;
  }
`;

const ShowPage: React.FC = () => {
  return (
    <PageContainer>
      <ShowDetails />
    </PageContainer>
  );
};

export default ShowPage;
