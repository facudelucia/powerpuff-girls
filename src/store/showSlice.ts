import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";
import { Show, Episode } from "../types/showTypes";

export interface ShowState {
  show: Show | null;
  episodes: Episode[];
  selectedEpisode: Episode | null;
  status: "idle" | "loading" | "failed";
}

const initialState: ShowState = {
  show: null,
  episodes: [],
  selectedEpisode: null,
  status: "idle",
};

export const fetchShowDetails = createAsyncThunk(
  "show/fetchShowDetails",
  async () => {
    const response = await axios.get("https://api.tvmaze.com/shows/6771");
    return response.data;
  }
);

export const fetchEpisodes = createAsyncThunk(
  "show/fetchEpisodes",
  async () => {
    const response = await axios.get(
      "https://api.tvmaze.com/shows/6771/episodes"
    );
    return response.data;
  }
);

const showSlice = createSlice({
  name: "show",
  initialState,
  reducers: {
    selectEpisode: (state, action) => {
      state.selectedEpisode =
        state.episodes.find(
          (episode: Episode) => episode.id === action.payload
        ) || null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchShowDetails.pending, (state) => {
        state.status = "loading";
      })
      .addCase(fetchShowDetails.fulfilled, (state, action) => {
        state.status = "idle";
        state.show = action.payload;
      })
      .addCase(fetchEpisodes.fulfilled, (state, action) => {
        state.episodes = action.payload;
      });
  },
});

export const { selectEpisode } = showSlice.actions;
export default showSlice.reducer;
