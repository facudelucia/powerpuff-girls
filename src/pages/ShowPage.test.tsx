import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import ShowPage from './ShowPage';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { store } from '../store';

describe('ShowPage Component', () => {
  test('renders ShowPage and shows loading state', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <ShowPage />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getByText(/loading/i)).toBeInTheDocument();
  });

  test('matches the snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <ShowPage />
        </MemoryRouter>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
