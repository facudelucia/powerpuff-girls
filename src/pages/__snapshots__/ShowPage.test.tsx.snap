// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowPage Component matches the snapshot 1`] = `
<div
  className="sc-iKTcqh lfjulL"
>
  <div>
    Loading...
  </div>
</div>
`;
